### DISCUSSION TITLE
REST API

### DISCUSSION THREAD

**Discussion by angrybrad** [craftcms/cms#8102] [Ideas]
> Created by: Tom (t@klepl.com) on 2015/04/20 17:54:59 +0000
> Votes at time of UserVoice import: 101

-----------------------------------

It would be great to have a way to access content stored in Craft for use in mobile apps and other types of front ends or systems, other than  fudging a JSON feed using templates.

Ideally there would be a section in the admin panel to define API endpoints and what channels they map to. Arguments for pagination and searching/filtering would be provided out of the box.

**Comment by angrybrad** [comment]
> Posted by Brandon Kelly (brandon@pixelandtonic.com) on 2015/03/06 03:54:45 +0000
-----------------------------------

There’s nothing stopping you from building custom REST APIs on top of Craft using custom plugin controllers. And there&#39;s at least one plugin out there that does it for you (https://github.com/wesrice/rest-easy).

We&#39;re not opposed to adding something into the core (Yii 2 actually makes it much easier, thanks to a base REST controller), but it would be quite a bit of work since we don&#39;t want to make any assumptions about the endpoint URLs, permissions, and how much data is sent to GET requests, etc., and coming up with a user-friendly UI for defining all that stuff will be a challenge.

**Comment by angrybrad** [comment]
> Posted by Jake Zerrer (jake@untapt.com) on 2015/07/24 18:06:51 +0000
-----------------------------------

Thinking about using Craft to back a very large commercial site. Craft&#39;s back-office editing system is amazing. Unfortunately, we can&#39;t use Craft unless content publishing and retrieval is available via a json REST API.

Any chance this is on your roadmap?

**Comment by angrybrad** [comment]
> Posted by Rob Erskine (rob@roberskine.com) on 2015/07/13 17:58:24 +0000
-----------------------------------

I find that I end up creating custom JSON routes to accomplish this for a lot of projects. Having it be done out of the box would be a huge timesaver. Even just using GETs to retrieve data would be awesome, however having the ability to POST with a client/secret/or another kind of auth would be amazing too.

**Comment by phoob** [comment]
Would you consider just going with GraphQL instead..? https://github.com/markhuot/craftql
